app/src/main/java/pl/futurecollars/invoicing/db/FileBasedDatabase.java

```diff
@@ -0,0 +1,113 @@
+package pl.futurecollars.invoicing.db;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import pl.futurecollars.invoicing.config.Configuration;
+import pl.futurecollars.invoicing.model.Invoice;
+import pl.futurecollars.invoicing.service.IdService;
+import pl.futurecollars.invoicing.service.JsonService;
+
+public class FileBasedDatabase implements Database {
+
+    private final File invoicesFile;
+    private final Path invoicesPath;
+    private final IdService idService;
+    private final JsonService jsonService = new JsonService();
+
+    public FileBasedDatabase() {
+        invoicesFile = new File(Configuration.INVOICES_PATH);
+        invoicesPath = Path.of(Configuration.INVOICES_PATH);
+        idService = new IdService(Path.of(Configuration.ID_PATH));
+    }
+
+    public FileBasedDatabase(File invoicesFile, Path invoicesPath, IdService idService) {
+        this.invoicesFile = invoicesFile;
+        this.invoicesPath = invoicesPath;
+        this.idService = idService;
+    }
+
+    @Override
+    public int save(Invoice invoice) {
+        int id = idService.getId();
+        invoice.setId(id);
+        idService.setId();
+
+        try {
+            Files.writeString(invoicesPath, jsonService.objectToString(invoice),
+                    invoicesFile.exists() ? StandardOpenOption.APPEND : StandardOpenOption.CREATE_NEW);
+            return id;
+        } catch (IOException exception) {
+            exception.printStackTrace();
+        }
+
+        return -1;
+    }
+
+    @Override
+    public Optional<Invoice> getById(int id) {
+        List<Invoice> filteredInvoices = getAll()
+                .stream()
+                .filter(invoice -> invoice.getId() == id)
+                .collect(Collectors.toList());
+
+        return filteredInvoices.isEmpty() ? Optional.empty() : Optional.ofNullable(filteredInvoices.get(0));
+    }
+
+    @Override
+    public List<Invoice> getAll() {
+        try {
+            return Files.readAllLines(invoicesPath)
+                    .stream()
+                    .map(jsonService::stringToObject)
+                    .collect(Collectors.toList());
+        } catch (IOException exception) {
+            exception.printStackTrace();
```

Comment:
jeśli chcesz coś logować to najlepiej byłoby użyć adnotacji lomboka (np. @Slf4j) i logować z jego pomocą - wtedy mamy większe możliwości zarządzania takimi logami (zapis do pliku, dzielenie na części, usuwanie co jakiś czas, itd), printStackTrace tylko wyrzuca do konsoli i nic z tym nie możemy zrobić dalej